feat(core): introduce resizable-card-layout component

DeepakSap14 opened the pull request:
#### Please provide a link to the associated issue.
fixes #4262 
#### Please provide a brief summary of this pull request.
This PR has implementation of initial version of Resizable-card-layout component.
https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?spaceKey=visualcore&title=OVP+Resizable+Layout

Pending feature:
1. Drag and drop
Issue link to follow: #4965 

**Dependency**:
Styles 17.0 release is required for the layout to work.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [x] Documentation Examples
- [x] Stackblitz works for all examples

InnaAtanasova commented:
A few "visual" problems that need to be fixed before reviewing the code:
1. I didn't see in your examples `Resize indication border` that is appearing on Card resize
2. I am not sure this is the correct focus for the Card:
<img width="355" alt="Screen Shot 2021-03-19 at 1 47 41 PM" src="https://user-images.githubusercontent.com/39598672/111822061-b9a72b80-88b9-11eb-82bb-07d39eecd9ca.png">
3. When you start resizing the Card, before ending the event, the Cards re-arrange as shown in the animation in the wiki doc that you provided in the template. The same animation is showing the Resize Indication border.

<img width="1466" alt="Screen Shot 2021-03-19 at 1 48 35 PM" src="https://user-images.githubusercontent.com/39598672/111823675-9aa99900-88bb-11eb-80a9-40dc6c43e7b2.png">

4. Your XL screen has 3 columns layout and in the specs it's specified that it should have 4. From the specs: "If the browser window width is between 1360px to 1679px, 4 columns of cards will be displayed." Maybe you should provide an example with 1679px screen and 4 columns for XL 
5. 
<img width="492" alt="Screen Shot 2021-03-19 at 1 57 28 PM" src="https://user-images.githubusercontent.com/39598672/111823218-135c2580-88bb-11eb-86ea-c7f9e17e0356.png">

The distance from the Header (B) and the first row of cards should be 1 rem. I saw in your examples you added vertical spacing on the Dialog body, but this is only on the documentation and only in case the user uses a dialog. This spacing should be provided on `fd-resizable-card-layout`

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.scss:
```diff
@@ -0,0 +1,23 @@
+}
+
+.fd-resizable-card-layout__item {
+    display: inline-flex !important;
+    max-width: 100%;
+    max-height: 100%;
+    margin: 0 !important;
+    padding: 0.5rem !important;
```
why are all rules important?

DeepakSap14 replied:
i was over-riding values coming from fundamental-styles.
but now using "resizable-card-layout.css" directly. this can be removed once styles are released and version upgraded.

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.scss:
```diff
@@ -0,0 +1,15 @@
+    position: relative;
+}
+
+.fd-resizable-card-layout--lg {
+    padding-left: 0.5rem !important;
+    padding-right: 0.5rem !important;
+}
+
+.fd-resizable-card-layout--xl {
+    padding-left: 2.5rem !important;
+    padding-right: 2.5rem !important;
+}
```
why do you need these?

DeepakSap14 replied:
i had missed xl screen size in fundamental styles.
now styles is corrected and using "resizable-card-layout.css" directly.

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,497 @@
+    layoutChange: EventEmitter<ResizableCardLayoutConfig> = new EventEmitter<ResizableCardLayoutConfig>();
+
+    /** @hidden */
+    // @ViewChildren(ResizableCardItemComponent)
+    // resizeCardItems: QueryList<ResizableCardItemComponent>;
```
remove unused code

DeepakSap14 replied:
removed

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,497 @@
+    // @ViewChildren(ResizableCardItemComponent)
+    // resizeCardItems: QueryList<ResizableCardItemComponent>;
+
+    /** @hidden */
+    @ContentChildren(ResizableCardItemComponent)
+    resizeCardItems: QueryList<ResizableCardItemComponent>;
+
+    /** @hidden */
+    layoutWidth: number;
+
+    /** @hidden */
+    layoutHeight: number;
```
all hidden props should be prefixed with `_`

DeepakSap14 replied:
fixed

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,497 @@
+    /** @hidden  */
+    private _destroy$ = new Subject<boolean>();
+
+    constructor(private readonly _changeDetectorRef: ChangeDetectorRef, private readonly _elementRef: ElementRef) {}
```
`_changeDetectorRef: ChangeDetectorRef` should be `_cd: ChangeDetectorRef` for consistency

DeepakSap14 replied:
changed to _cd

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,484 @@
+    private _resizeDirection: ResizeDirection;
+    private _maxCardWidth: number;
+
+    constructor(private readonly _changeDetectorRef: ChangeDetectorRef, private readonly _elementRef: ElementRef) {}
```
_cd: ChangeDetectorRef

DeepakSap14 replied:
changed to _cd

InnaAtanasova commented:
<img width="2273" alt="Screen Shot 2021-03-19 at 2 15 12 PM" src="https://user-images.githubusercontent.com/39598672/111825199-91213080-88bd-11eb-94f6-56e4671aef5a.png">

Problems with content overflow and focus: see cards 6 and 7

InnaAtanasova commented:
It's missing examples and explanation how to make a layout non-resizable. Is it the whole layout or specific Cards can be disabled?

JKMarkowski commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,484 @@
+        }
+
+        // increase/decrease width of card in order of 20rem
+        if (Math.abs(this.cardWidth - this._prevCardWidth) > 0) {
+            this._horizontalStepResizing();
+        }
+
+        // increase/decrease height of card in order of 1rem
+        if (Math.abs(this.cardHeight - this._prevCardHeight) > 0) {
+            this._verticalStepResizing();
+        }
```
Could you do some offset, just to not let user resize cards by default, for example user needs to resize by 20px to change width/height?

![Peek 2021-03-23 12-04](https://user-images.githubusercontent.com/26483208/112136930-ed42c800-8bcf-11eb-87e0-dec9140f3e9f.gif)

DeepakSap14 replied:
added horizontalOffset of 20px and verticalOffset of 10px. now it works fine.

JKMarkowski commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,484 @@
+
+    /** @hidden */
+    ngOnDestroy(): void {
+        this._resizeDebounce$.next();
+        this._resizeDebounce$.complete();
```
Should this event be thrown on destroy?

DeepakSap14 replied:
removed .next()

JKMarkowski commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.scss:
```diff
@@ -0,0 +1,15 @@
+}
+
+.fd-resizable-card-layout--lg {
+    padding-left: 0.5rem !important;
+    padding-right: 0.5rem !important;
+}
+
+.fd-resizable-card-layout--xl {
+    padding-left: 2.5rem !important;
+    padding-right: 2.5rem !important;
+}
```
Let's not use important anywhere, where it's not needed. You can put stronger selector

DeepakSap14 replied:
i was over-riding values coming from fundamental-styles.
but now using "resizable-card-layout.css" directly. this can be removed once styles are released and version upgraded.

JKMarkowski reviewed (changes requested):
Hello @DeepakSap14 Good job! I adde dsome comment. Could you rebase this branch to newest `main`? I think there is something wrong with card focus - header should be focused, instead of whole card. Like there 
![Peek 2021-03-23 12-14](https://user-images.githubusercontent.com/26483208/112138039-524aed80-8bd1-11eb-9bb3-de5fc9dbda69.gif)

DeepakSap14 commented:
> A few "visual" problems that need to be fixed before reviewing the code:
> 
> 1. I didn't see in your examples `Resize indication border` that is appearing on Card resize
> 2. I am not sure this is the correct focus for the Card:
> 
> <img alt="Screen Shot 2021-03-19 at 1 47 41 PM" width="355" src="https://user-images.githubusercontent.com/39598672/111822061-b9a72b80-88b9-11eb-82bb-07d39eecd9ca.png">
> 
> 3. When you start resizing the Card, before ending the event, the Cards re-arrange as shown in the animation in the wiki doc that you provided in the template. The same animation is showing the Resize Indication border.
> <img alt="Screen Shot 2021-03-19 at 1 48 35 PM" width="1466" src="https://user-images.githubusercontent.com/39598672/111823675-9aa99900-88bb-11eb-80a9-40dc6c43e7b2.png">
> 
> 1. Your XL screen has 3 columns layout and in the specs it's specified that it should have 4. From the specs: "If the browser window width is between 1360px to 1679px, 4 columns of cards will be displayed." Maybe you should provide an example with 1679px screen and 4 columns for XL
> 
> <img alt="Screen Shot 2021-03-19 at 1 57 28 PM" width="492" src="https://user-images.githubusercontent.com/39598672/111823218-135c2580-88bb-11eb-86ea-c7f9e17e0356.png">
> 
> The distance from the Header (B) and the first row of cards should be 1 rem. I saw in your examples you added vertical spacing on the Dialog body, but this is only on the documentation and only in case the user uses a dialog. This spacing should be provided on `fd-resizable-card-layout`

1. added resize indication border. but expanding of border will be implemented while fixing peding issue #4965.
2. corrected focus now.
3. Animation of card shifting will be done with issue #4965
4. XL screen is showing 4 columns, some of cards are taking 2 or more columns.
5. added space between Header and first row of cards.

DeepakSap14 commented:
> <img alt="Screen Shot 2021-03-19 at 2 15 12 PM" width="2000" src="https://user-images.githubusercontent.com/39598672/111825199-91213080-88bd-11eb-94f6-56e4671aef5a.png">
> 
> Problems with content overflow and focus: see cards 6 and 7

1. This is because card has content which is not fitting in current dimension of card provided. for these cases data need to be increased/decreased according to the dimension available. this i will implement while implementing issue #4965.
This is documentation side change only.

DeepakSap14 commented:
> It's missing examples and explanation how to make a layout non-resizable. Is it the whole layout or specific Cards can be disabled?

1. Now added example for Non-resizable card in first example, where card-2 is not resizable.

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+        // resizing card will go over other cards
+        this.zIndex = 1;
+
+        if (this._resizeDirection === 'both') {
```
please use switch here

DeepakSap14 replied:
fixed, using switch case now.

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+export type ResizeDirection = 'vertical' | 'horizontal' | 'both';
+
+// Card resizes in step of fixed values. values are in pixel
+export const HorizontalResizeStep = 320;
```
```suggestion
export const horizontalResizeStep = 320;
```

DeepakSap14 replied:
renamed.

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+        // resizing card will go over other cards
+        this.zIndex = 1;
+
+        if (this._resizeDirection === 'both') {
+            this._horizontalResizing(event.clientX);
+            this._verticalResizing(event.clientY);
+        } else if (this._resizeDirection === 'horizontal') {
+            this._horizontalResizing(event.clientX);
+        } else {
+            this._verticalResizing(event.clientY);
+        }
```
```suggestion
        const { clientX, clientY } = event;
        if (this._resizeDirection === 'both') {
            this._horizontalResizing(clientX);
            this._verticalResizing(clientY);
        } else if (this._resizeDirection === 'horizontal') {
            this._horizontalResizing(clientX);
        } else {
            this._verticalResizing(clientY);
        }
```

DeepakSap14 replied:
fixed

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+    /** Sets focus on the element */
+    focus(): void {
+        const header = this._elementRef.nativeElement.querySelector('.fd-card__header');
+        if (header) {
+            header.focus();
+        }
```
```suggestion
        header?.focus();
```

DeepakSap14 replied:
fixed

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,497 @@
+
+    /** @hidden Sets number of column in layout, based on LayoutSize passed */
+    private _setLayoutColumns(layoutSize: LayoutSize): void {
+        switch (layoutSize) {
+            case 'sm':
+                this._columns = 1;
+                this._paddingLeft = 8;
+                this.layoutWidth = this._columns * HorizontalResizeStep + 2 * this._paddingLeft;
+                break;
+            case 'md':
+                this._columns = 2;
+                this._paddingLeft = 16;
+                this.layoutWidth = this._columns * HorizontalResizeStep + 2 * this._paddingLeft + gap;
+                break;
+            case 'lg':
+                this._columns = 3;
+                this._paddingLeft = 16;
+                this.layoutWidth = this._columns * HorizontalResizeStep + 2 * this._paddingLeft + 2 * gap;
+                break;
+            case 'xl':
+                this._columns = 4;
+                this._paddingLeft = 48;
+                this.layoutWidth = this._columns * HorizontalResizeStep + 2 * this._paddingLeft + 3 * gap;
+                break;
+            default:
+                this._columns = 4;
+                this._paddingLeft = 48;
+                this.layoutWidth = this._columns * HorizontalResizeStep + 2 * this._paddingLeft + 3 * gap;
+        }
```
```suggestion
        const layoutWidthDefault = this._columns * HorizontalResizeStep + 2 * this._paddingLeft;
        switch (layoutSize) {
            case 'sm':
                this._columns = 1;
                this._paddingLeft = 8;
                this.layoutWidth = layoutWidthDefault;
                break;
            case 'md':
                this._columns = 2;
                this._paddingLeft = 16;
                this.layoutWidth = layoutWidthDefault + gap;
                break;
            case 'lg':
                this._columns = 3;
                this._paddingLeft = 16;
                this.layoutWidth = layoutWidthDefault + 2 * gap;
                break;
            case 'xl':
                this._columns = 4;
                this._paddingLeft = 48;
                this.layoutWidth = layoutWidthDefault + 3 * gap;
                break;
            default:
                this._columns = 4;
                this._paddingLeft = 48;
                this.layoutWidth = layoutWidthDefault + 3 * gap;
        }
```

DeepakSap14 replied:
fixed.
have to keep `const layoutWidthDefault = this._columns * HorizontalResizeStep + 2 * this._paddingLeft;` inside case statement; as this._columns and this._paddingLeft; will not be available outside.

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,497 @@
+     * @hidden try to set card at given height
+     */
+    private _isPositionSetSuccess(height: number, card: ResizableCardItemComponent): boolean {
+        const columnPositions = new Array();
```
Why not use [] instead of new Array?

DeepakSap14 replied:
fixed

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,497 @@
+    private _getSortedUniqueHeights(): number[] {
+        const tempArray = this._columnsHeight.slice();
+        const sortedColumnsHeightArray = tempArray.sort(comparer);
+        const uniqueHeights = new Array();
```
the same

DeepakSap14 replied:
fixed

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,497 @@
+
+        // try to set towards left from available card position
+        // eg. [1, 2, 3, 4] columnsPositions
+        for (const columnPosition of columnPositions) {
+            if (!isFitting) {
```
```suggestion
        for (const columnPosition of columnPositions) {
            if (isFitting) {
                break;
            }
```

Can we do like that to decrease nesting?

DeepakSap14 replied:
fixed

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,497 @@
+        const cardColSpan = Math.floor(card.cardWidth / HorizontalResizeStep);
+
+        for (const columnPosition of columnPositions) {
+            if (!isFitting) {
```
same here

DeepakSap14 replied:
fixed

KevinOkamoto commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.scss:
```diff
@@ -0,0 +1,2255 @@
+/*!
+ * Fundamental Library Styles v0.17.0-rc.30
+ * Copyright (c) 2021 SAP SE or an SAP affiliate company.
+ * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/main/LICENSE)
+ */
+ @charset "UTF-8";
+ [class*=sap-icon] {
```
Why do you need to include the icon CSS here? You should use `@import` instead.

DeepakSap14 replied:
Resizable-card-layout css is not yet released.
as per core team suggestion, i pasted dist/resizable-card-layout.css here.
later when it is released; this can be replaced.

KevinOkamoto replied:
But why are you copying the `sap-icon` CSS? Doesn't that already exist in the `icon.css` - it should be imported in the `resizable-card-layout.scss` too.

droshev replied:
@DeepakSap14 ngx is on top of the latest styles, please, clean up this

KevinOkamoto commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.scss:
```diff
@@ -0,0 +1,2255 @@
+/*!
+ * Fundamental Library Styles v0.17.0-rc.30
+ * Copyright (c) 2021 SAP SE or an SAP affiliate company.
+ * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/main/LICENSE)
+ */
+ @charset "UTF-8";
+ [class*=sap-icon] {
+   font-size: .875rem;
```
Should `@import` the fundamental-styles icon.css here.

DeepakSap14 replied:
same as above

JKMarkowski replied:
There is no `resizable-card.css` in `0.17.0`, let's keep it as it is

KevinOkamoto commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+    @Input()
+    title: string;
+
+    /** Position rank  of card */
+    @Input()
+    rank: number = cardRank++;
```
If user sets the `rank` on another card, `cardRank` may not always be the maximum value. You might want to use setter-getter here and update `cardRank` to be the maximum value whenever  `rank` is set.

DeepakSap14 replied:
fixed now

DeepakSap14 replied:
fixed now

KevinOkamoto reviewed (changes requested):
See comments

KevinOkamoto reviewed (changes requested):
When a card is not resizable, we should not show the corner expand icon on hover.

<img width="346" alt="Screen Shot 2021-03-25 at 7 40 14 AM" src="https://user-images.githubusercontent.com/48103491/112491567-ac68c180-8d3d-11eb-80e3-c2d7b0094400.png">

DeepakSap14 commented:
> When a card is not resizable, we should not show the corner expand icon on hover.
> 
> <img alt="Screen Shot 2021-03-25 at 7 40 14 AM" width="346" src="https://user-images.githubusercontent.com/48103491/112491567-ac68c180-8d3d-11eb-80e3-c2d7b0094400.png">

This is fixed now.

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,604 @@
+        }
+
+        // Emit resizing event on some interval. improves performance.
+        this._resizeDebounce$.next(new PositionChange(this._prevX - event.clientX, this._prevY - event.clientY));
```
```suggestion
        this._resizeDebounce$.next(new PositionChange(this._prevX - clientX, this._prevY - clientY));
```

DeepakSap14 replied:
fixed

Betrozov commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,604 @@
+            this.stepChange.emit(this._getResizedEventObject());
+        }
+
+        this._prevX = event.clientX;
+        this._prevY = event.clientY;
```
```suggestion
        this._prevX = clientX;
        this._prevY = clientY;
```

DeepakSap14 replied:
fixed

JKMarkowski commented:
I rechecked code - looks good - incredible job! I some behaviour glitches for resizing - could you check that again?
![resizablecard1](https://user-images.githubusercontent.com/26483208/112617599-73a41780-8e25-11eb-8950-33392ca5c8b4.gif)

DeepakSap14 commented:
> Hello @DeepakSap14
> I rechecked code - looks good - incredible job! I some behaviour glitches for resizing - could you check that again?
> ![resizablecard1](https://user-images.githubusercontent.com/26483208/112617599-73a41780-8e25-11eb-8950-33392ca5c8b4.gif)

This is due to that resize is more than 2 columns even by some pixels. In this case i am not checking offset.
i checked offset only when initially starting to resize.
Do you want to have offset here as well?

InnaAtanasova commented:
Is it supposed to be like this?
<img width="1650" alt="Screen Shot 2021-03-30 at 4 00 01 PM" src="https://user-images.githubusercontent.com/39598672/113049578-cce7b000-9171-11eb-9107-4fea3f6bb493.png">

InnaAtanasova commented:
When you start resizing a certain card, the other cards' positions are updated before releasing the card. The behaviour is shown in the wiki page linked in the issue template. Currently, when you start resizing a card, the other cards do not move:
<img width="1380" alt="Screen Shot 2021-03-30 at 4 04 20 PM" src="https://user-images.githubusercontent.com/39598672/113050032-60b97c00-9172-11eb-8a4d-66755a0cd9ff.png">
<img width="1391" alt="Screen Shot 2021-03-30 at 4 03 39 PM" src="https://user-images.githubusercontent.com/39598672/113050035-61eaa900-9172-11eb-8e0c-3224de39d3f1.png">


Expected behaviour:
1. Starting the resize:
<img width="630" alt="1" src="https://user-images.githubusercontent.com/39598672/113050405-cd347b00-9172-11eb-91e9-dc4003a07cff.png">

2. While resizing:
<img width="529" alt="2" src="https://user-images.githubusercontent.com/39598672/113050446-d6254c80-9172-11eb-8094-05ffe81cfa0e.png">

DeepakSap14 commented:
> Hello @DeepakSap14
> I rechecked code - looks good - incredible job! I some behaviour glitches for resizing - could you check that again?
> ![resizablecard1](https://user-images.githubusercontent.com/26483208/112617599-73a41780-8e25-11eb-8950-33392ca5c8b4.gif)

@InnaAtanasova This behaviour is fixed now.

DeepakSap14 commented:
> Is it supposed to be like this?
> <img alt="Screen Shot 2021-03-30 at 4 00 01 PM" width="1650" src="https://user-images.githubusercontent.com/39598672/113049578-cce7b000-9171-11eb-9107-4fea3f6bb493.png">
> 
> <img alt="Screen Shot 2021-03-30 at 3 59 48 PM" width="1670" src="https://user-images.githubusercontent.com/39598672/113049673-eb4dab80-9171-11eb-9784-77d54eb78d67.png">

@InnaAtanasova This is fixed now, not able to re-produce this now.

DeepakSap14 commented:
> When you start resizing a certain card, the other cards' positions are updated before releasing the card. The behaviour is shown in the wiki page linked in the issue template. Currently, when you start resizing a card, the other cards do not move:
> <img alt="Screen Shot 2021-03-30 at 4 04 20 PM" width="1380" src="https://user-images.githubusercontent.com/39598672/113050032-60b97c00-9172-11eb-8a4d-66755a0cd9ff.png">
> <img alt="Screen Shot 2021-03-30 at 4 03 39 PM" width="1391" src="https://user-images.githubusercontent.com/39598672/113050035-61eaa900-9172-11eb-8e0c-3224de39d3f1.png">
> 
> Expected behaviour:
> 
> 1. Starting the resize:
> 
> <img alt="1" width="630" src="https://user-images.githubusercontent.com/39598672/113050405-cd347b00-9172-11eb-91e9-dc4003a07cff.png">
> 
> 1. While resizing:
> 
> <img alt="2" width="529" src="https://user-images.githubusercontent.com/39598672/113050446-d6254c80-9172-11eb-8094-05ffe81cfa0e.png">

@InnaAtanasova , i have raised issue #4965 to track this. I will take this as an enhancement.

droshev reviewed (changes requested):
See my comments

DeepakSap14 commented:
Moved to WIP.
Need to implement Animation behaviour of card while resizing of a card.
It will close #4965 .

DeepakSap14 commented:
@InnaAtanasova @droshev 
This component can be reviewed now.
Animation behaviour is implemented.
because of using absolute positions for cards, implementing Drag-and-drop is challenging and i did not find any solution yet.
i will address that in another PR, it will not introduce any breaking change.

InnaAtanasova commented:
Lots of changes have been introduced since the reviewers approved, you need to re-request review. 
You need to add a label next to the component title in the documentation that this component is not finished yet. You can add `experimental` Info Label, for example.

Also, there's an issue with the Cards overflow, mostly noticeable in the small screens examples. The bottom border is not visible and the table cells overflow: 

<img width="810" alt="Screen Shot 2021-04-29 at 12 20 34 PM" src="https://user-images.githubusercontent.com/39598672/116585397-5b249280-a8e6-11eb-8179-ba6eea93c3ef.png">
<img width="583" alt="Screen Shot 2021-04-29 at 12 20 42 PM" src="https://user-images.githubusercontent.com/39598672/116585399-5bbd2900-a8e6-11eb-87b8-567b77252ad2.png">

KevinOkamoto commented:
It looks good. If you can address @InnaAtanasova 's issue with the card overflow, I will approve it.

sKudum commented:
Hi @DeepakSap14 ,

Few queries, rest looks good to me..approving it.

1).On tab items border is blue, is it acceptable? Is user not allowed to navigate items by keyboard actions.

<img width="361" alt="2021-04-30_16-25-44" src="https://user-images.githubusercontent.com/53487468/116686420-66a0b800-a9d1-11eb-9925-44d262dbb143.png">


2). Content is overflowing from card.

<img width="346" alt="2021-04-30_15-59-17" src="https://user-images.githubusercontent.com/53487468/116686518-8afc9480-a9d1-11eb-9c2c-e6bc1d893a57.png">


3).On more item display, left over card space is varying from item to item, is it expected behaviour?


![2021-04-30_16-39-43 (1)](https://user-images.githubusercontent.com/53487468/116687439-cea3ce00-a9d2-11eb-91dd-ae6a6d43f3e4.gif)

DeepakSap14 commented:
> Hi @DeepakSap14 ,
> 
> Few queries, rest looks good to me..approving it.
> 
> 1).On tab items border is blue, is it acceptable? Is user not allowed to navigate items by keyboard actions.
> 
> <img alt="2021-04-30_16-25-44" width="361" src="https://user-images.githubusercontent.com/53487468/116686420-66a0b800-a9d1-11eb-9925-44d262dbb143.png">
> 
> 2). Content is overflowing from card.
> 
> <img alt="2021-04-30_15-59-17" width="346" src="https://user-images.githubusercontent.com/53487468/116686518-8afc9480-a9d1-11eb-9c2c-e6bc1d893a57.png">
> 
> 3).On more item display, left over card space is varying from item to item, is it expected behaviour?
> 
> ![2021-04-30_16-39-43 (1)](https://user-images.githubusercontent.com/53487468/116687439-cea3ce00-a9d2-11eb-91dd-ae6a6d43f3e4.gif)

1. blue border is default browser behaviour. same is happening for list component as well.
<img width="1441" alt="core-list" src="https://user-images.githubusercontent.com/57661487/116689051-35c28200-a9d5-11eb-9f1e-c3d348a8cd5f.png">
Keyborad navigation of List items should happen, i will check that.
2. It is fixed now. overflow should not happen now. 
3. if data is not available any more and card is stretched, white space will be shown. As much you stretch white space will be shown. It is expected.

InnaAtanasova reviewed (commented):
@DeepakSap14 how is the height of the Cards determined? Why are they different when the content is the same?
<img width="247" alt="Screen Shot 2021-05-03 at 4 54 09 PM" src="https://user-images.githubusercontent.com/39598672/116932891-d660ae00-ac30-11eb-85d2-d262d01a1d11.png">

DeepakSap14 commented:
> @DeepakSap14 how is the height of the Cards determined? Why are they different when the content is the same?
> <img alt="Screen Shot 2021-05-03 at 4 54 09 PM" width="247" src="https://user-images.githubusercontent.com/39598672/116932891-d660ae00-ac30-11eb-85d2-d262d01a1d11.png">

Hi @InnaAtanasova,
Height of the card is determined by property value `cardHeightRowSpan`.
According to the height given, application data should be able to determine, how much data to show. If data given is quite less than the height given or when card resized; but more data is not available. then it will be showing white space.
https://ux.wdf.sap.corp/fiori-design-web/resizable-card-layout-overview-page/#dealing-with-white-space

In my first example, i have shown the data passing dynamically based on card height and width. Here more data is shown when card height increases.

InnaAtanasova commented on apps/docs/src/app/core/component-docs/resizable-card-layout/examples/resizable-card-layout-example.component.scss:
```diff
@@ -0,0 +1,8 @@
+.fd-docs-cursor {
+    cursor: auto !important;
```
why do you need to change the cursor? This is a documentation page, you only add styling for representational purposes

DeepakSap14 replied:
removed now. because of using popover, cursor changing to "grab".

InnaAtanasova commented on apps/docs/src/app/core/component-docs/resizable-card-layout/examples/resizable-card-layout-example-itemconfig.component.scss:
```diff
@@ -0,0 +1,8 @@
+.fd-docs-cursor {
+    cursor: auto !important;
```
same as the other comment for cursor

DeepakSap14 replied:
same as above

InnaAtanasova commented on apps/docs/src/app/core/component-docs/resizable-card-layout/examples/resizable-card-layout-example-layoutconfig.component.scss:
```diff
@@ -0,0 +1,8 @@
+.fd-docs-cursor {
+    cursor: auto !important;
```
same as the other comment for cursor

InnaAtanasova commented on apps/docs/src/app/core/component-docs/resizable-card-layout/examples/resizable-card-layout-example.component.scss:
```diff
@@ -0,0 +1,8 @@
+    cursor: auto !important;
+}
+
+.fd-docs-center-align {
+    display: flex;
+    justify-content: center;
+}
```
this is repeated 3 times. You can make a single scss file and import it in the 3 files, no need to create a separate file for each component. Scss files can be reused

DeepakSap14 replied:
using one scss file now. removed others.

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.scss:
```diff
@@ -0,0 +1,10 @@
+@import "~fundamental-styles/dist/resizable-card-layout";
+
+.fd-resizable-card-expansion {
+    transition-duration: 350ms;
+    animation-timing-function: ease-in-out;
+}
+
+.fd-resizable-card-border-display {
+    display: flex;
+}
```
does not belong to fund-ngx

DeepakSap14 replied:
currently i have not got animation duration values from designer. I had asked them about couple of weeks ago some other questions, but this component is not in their priority list. and they are not able to provide the answer.
I have asked them again today. whenever answer comes, we can move these values to styles.
hope this is fine to unlock the development. i will keep an issue open for this.

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.scss:
```diff
@@ -0,0 +1,10 @@
+@import "~fundamental-styles/dist/resizable-card-layout";
+
+.fd-resizable-card-expansion {
+    transition-duration: 350ms;
+    animation-timing-function: ease-in-out;
```
are these values taken from the specs?

DeepakSap14 replied:
these values are not provided in specs. have asked designers. but currently they do not have these values and this component is not in their priority for now. for unblocking the development  i have used these values, which come close to achieving the required behaviour. 
i have asked for the values again. if answers come, i will update it.

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.html:
```diff
@@ -0,0 +1,33 @@
+    <div class="fd-resizable-card-expansion" [style.width.px]="resizeIndicationBorderWidth" [style.height.px]="cardHeight"></div>
+</div>
+
+<ng-template #verticalHandleTpt>
```
Does `Tpt` stand for `template`? Please rename with more meaningful name.

DeepakSap14 replied:
corrected.

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+        return this._config;
+    }
+
+    set config(config: ResizableCardItemConfig) {
```
missing typedocs

DeepakSap14 replied:
added typedoc

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+export const verticalResizeOffset = 10;
+export type CardState = 1 | 0 | -1;
+
+let cardRank = 1;
```
please add a comment about `cardRank` - what it is and what value it accepts.

DeepakSap14 replied:
added typedoc

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+    get rank(): number {
+        return this._rank;
+    }
+    set rank(rankValue: number) {
```
missing typedocs for setter

DeepakSap14 replied:
added typedoc

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+    get cardHeightRowSpan(): number {
+        return this._cardHeightRowSpan;
+    }
+    set cardHeightRowSpan(rowSpan: number) {
```
missing typedocs for setter

DeepakSap14 replied:
added typedoc

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+    get cardWidthColSpan(): number {
+        return this._cardWidthColSpan;
+    }
+    set cardWidthColSpan(colSpan: number) {
```
missing typedocs for setter

DeepakSap14 replied:
added typedoc

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+    get cardMiniHeaderRowSpan(): number {
+        return this._cardMiniHeaderRowSpan;
+    }
+    set cardMiniHeaderRowSpan(rowSpan: number) {
```
missing typedocs for setter

DeepakSap14 replied:
added typedoc

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+    get cardMiniContentRowSpan(): number {
+        return this._cardMiniContentRowSpan;
+    }
+    set cardMiniContentRowSpan(rowSpan: number) {
```
missing typedocs for setter

DeepakSap14 replied:
added typedoc

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+    /** Change in rank from previous value*/
+    prevRank = 0;
+
+    private _cardWidthColSpan: number;
+    private _cardHeightRowSpan: number;
+    private _cardMiniHeaderRowSpan: number;
+    private _cardMiniContentRowSpan?: number;
```
all these should have @hidden

DeepakSap14 replied:
added @hidden

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+
+    /** @hidden config values for card */
+    private _config: ResizableCardItemConfig;
+    private _prevX: number;
+    private _prevY: number;
+    private _prevCardWidth: number;
+    private _prevCardHeight: number;
```
all these should have @hidden

DeepakSap14 replied:
added @hidden

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+
+    /** @hidden flag to control resize */
+    private _resize = false;
+    private _resizeDirection: ResizeDirection;
+    private _maxColumn: number;
+    private _rank: number;
```
all these should have @hidden

DeepakSap14 replied:
added @hidden

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,588 @@
+                this._horizontalResizing(clientX);
+                break;
+            case 'vertical':
+                this._verticalResizing(clientY);
```
no default?

DeepakSap14 replied:
added default.

InnaAtanasova commented on libs/core/src/lib/resizable-card-layout/resizable-card-layout.component.ts:
```diff
@@ -0,0 +1,553 @@
+        return this._layout;
+    }
+
+    set layoutSize(layoutSize: LayoutSize) {
```
typedocs

DeepakSap14 replied:
added typedoc

mikerodonnell89 reviewed (changes requested):
Is the entire page supposed to have the grab cursor? Shouldn't it just be on the cards themselves?  Also I'm seeing some weird behavior with using the trackpad on my mac, i can only seem to adjust the size of the card once, then clicking and dragging does nothing

DeepakSap14 commented:
> Is the entire page supposed to have the grab cursor? Shouldn't it just be on the cards themselves? Also I'm seeing some weird behavior with using the trackpad on my mac, i can only seem to adjust the size of the card once, then clicking and dragging does nothing

Hi @mikerodonnell89 
1. as i am using popover to open example in full screen, grab cursor is coming by default. i had removed that behaviour but was later asked to remove it here in review comments. https://github.com/SAP/fundamental-ngx/pull/4964#discussion_r626842662

2. i am able to resize card any number of time.
![2021-05-11_09-20-01 (1)](https://user-images.githubusercontent.com/57661487/117755923-cea99680-b23a-11eb-86ac-ac99a109701a.gif)

hanjunt commented:
Hello, I experience the same issue as Mike. This took place under Resizable Card Layout - Medium Screen
https://user-images.githubusercontent.com/46663585/117999820-3d5a2300-b313-11eb-91fe-4dd59de0e107.mov

DeepakSap14 commented:
@mikerodonnell89 @hanjunt @InnaAtanasova 
resizing issue at mini-content is fixed now.
animation duration values are updated as per fiori.
It is looking Good now, please review again.

droshev commented on apps/docs/src/app/core/component-docs/resizable-card-layout/resizable-card-layout-docs-header/resizable-card-layout-docs-header.component.html:
```diff
@@ -0,0 +1,11 @@
+    <fd-info-label label="Experimental" title="development in-progress"></fd-info-label>
+</header>
+<description>
+    Resizable Card Layout enables users to define a personalized card layout by changing size of the card as well as position of the card.
```
how can we change the position? is it through DnD or through the API?

DeepakSap14 replied:
documented

droshev commented on apps/docs/src/app/core/component-docs/resizable-card-layout/examples/resizable-card-layout-example.component.ts:
```diff
@@ -0,0 +1,281 @@
+import { Component, TemplateRef } from '@angular/core';
```
here the example has 3 files but on the doc we see only the html one:
![Screen Shot 2021-05-17 at 6 03 07 AM](https://user-images.githubusercontent.com/4380815/118471367-af48b880-b6d5-11eb-9826-71efa8c8c4aa.png)
![Screen Shot 2021-05-17 at 6 03 45 AM](https://user-images.githubusercontent.com/4380815/118471368-afe14f00-b6d5-11eb-8e5a-fc097eb2567a.png)


Same for the other 2 examples

DeepakSap14 replied:
fixed, now showing all 3 files.

droshev commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,656 @@
+
+    // return rank of card
+    @Input()
+    get rank(): number {
```
Can we be more verbose what rank is?

DeepakSap14 replied:
added more context for rank

droshev commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,656 @@
+        this._setCardHeight();
+    }
+
+    /** return number of row span card mini header height takes */
```
what is mini header?

DeepakSap14 replied:
added more context for mini-header

droshev commented on libs/core/src/lib/resizable-card-layout/resizable-card-item/resizable-card-item.component.ts:
```diff
@@ -0,0 +1,656 @@
+
+    /** return number of row span card mini content height takes */
+    @Input()
+    get cardMiniContentRowSpan(): number {
```
what is mini content?

DeepakSap14 replied:
added more context for mini-content

droshev reviewed (changes requested):
We need to be more verbose documenting such big layout/feature. The API would need to be also more documented

DeepakSap14 commented on apps/docs/src/app/core/component-docs/resizable-card-layout/resizable-card-layout-docs.component.html:
```diff
@@ -0,0 +1,37 @@
+<fd-docs-section-title id="resizable-card-layout-default" componentName="resizable-card-layout">
+    Resizable Card Layout
```
have documented more about the layout and its features.

DeepakSap14 commented:
@InnaAtanasova @droshev @kari-leg @mikerodonnell89 
Can this be approved and merged?
Do we have any blocker?

Seamoo13 commented:
- [ ] card contents can exceed card boundary (resizeable card layout example)
![2021-05-20_7](https://user-images.githubusercontent.com/47522152/118981372-f27b8380-b982-11eb-882e-ae313cdac3c6.png)

- [ ] cards can overlap (resizeable card layout example)
![2021-05-20_6](https://user-images.githubusercontent.com/47522152/118981441-00c99f80-b983-11eb-8398-90245721aadd.png)
![2021-05-20_5](https://user-images.githubusercontent.com/47522152/118981443-01623600-b983-11eb-8478-d63f543ba5e2.png)
https://user-images.githubusercontent.com/47522152/118981542-1b9c1400-b983-11eb-9698-28e8ffe9802b.mp4

- [ ] for Resizable Card Layout using Layout config and item config examples, when you make the card small, the card items exceeds the card space and items not collapsed/hidden
![2021-05-20_8](https://user-images.githubusercontent.com/47522152/118981985-a250f100-b983-11eb-8bee-53e250488066.png)
![2021-05-20_9](https://user-images.githubusercontent.com/47522152/118982732-610d1100-b984-11eb-90f3-d868344f3201.png)
https://user-images.githubusercontent.com/47522152/118982752-65392e80-b984-11eb-8d55-9e66a4d8b987.mp4
https://user-images.githubusercontent.com/47522152/118983197-d7117800-b984-11eb-8d34-1d3c3ba75e98.mp4

DeepakSap14 commented:
> * [x]  card contents can exceed card boundary (resizeable card layout example)
>   ![2021-05-20_7](https://user-images.githubusercontent.com/47522152/118981372-f27b8380-b982-11eb-882e-ae313cdac3c6.png)
> * [x]  cards can overlap (resizeable card layout example)
>   ![2021-05-20_6](https://user-images.githubusercontent.com/47522152/118981441-00c99f80-b983-11eb-8398-90245721aadd.png)
>   ![2021-05-20_5](https://user-images.githubusercontent.com/47522152/118981443-01623600-b983-11eb-8478-d63f543ba5e2.png)
>   https://user-images.githubusercontent.com/47522152/118981542-1b9c1400-b983-11eb-9698-28e8ffe9802b.mp4
> * [x]  for Resizable Card Layout using Layout config and item config examples, when you make the card small, the card items exceeds the card space and items not collapsed/hidden
>   ![2021-05-20_8](https://user-images.githubusercontent.com/47522152/118981985-a250f100-b983-11eb-8bee-53e250488066.png)
>   ![2021-05-20_9](https://user-images.githubusercontent.com/47522152/118982732-610d1100-b984-11eb-90f3-d868344f3201.png)
>   https://user-images.githubusercontent.com/47522152/118982752-65392e80-b984-11eb-8d55-9e66a4d8b987.mp4
>   https://user-images.githubusercontent.com/47522152/118983197-d7117800-b984-11eb-8d34-1d3c3ba75e98.mp4

@Seamoo13 , these has been fixed now.
please verify.

Seamoo13 commented:
still able to make the content exceed card space when making small adjustments to the card height
(can see this instantly by opening large or xl screen on Resizable Card Layout using Item config example and looking at the 1st card - card botom border not seen)
![2021-05-21_2](https://user-images.githubusercontent.com/47522152/119119885-527e3280-ba34-11eb-935d-33634cb9b499.png)
![2021-05-21_3](https://user-images.githubusercontent.com/47522152/119120021-76da0f00-ba34-11eb-82fd-b17ac97431bc.png)
![2021-05-21_4](https://user-images.githubusercontent.com/47522152/119120202-b274d900-ba34-11eb-9e02-4e2ae1725e22.png)

DeepakSap14 commented:
> still able to make the content exceed card space when making small adjustments to the card height
> (can see this instantly by opening large or xl screen on Resizable Card Layout using Item config example and looking at the 1st card - card botom border not seen)
> ![2021-05-21_2](https://user-images.githubusercontent.com/47522152/119119885-527e3280-ba34-11eb-935d-33634cb9b499.png)
> ![2021-05-21_3](https://user-images.githubusercontent.com/47522152/119120021-76da0f00-ba34-11eb-82fd-b17ac97431bc.png)
> ![2021-05-21_4](https://user-images.githubusercontent.com/47522152/119120202-b274d900-ba34-11eb-9e02-4e2ae1725e22.png)

it is fixed now. A corner case, thanks for pointing out.

InnaAtanasova commented:
Why the cursor is `grab`?

InnaAtanasova commented:
<img width="1083" alt="Screen Shot 2021-05-25 at 2 47 41 PM" src="https://user-images.githubusercontent.com/39598672/119551958-30541f80-bd68-11eb-8935-5b60dd4095e6.png">

Shouldn't the layout be responsive? The cards should move to adjust the new screen size. Am I missing something?

mikerodonnell89 commented:
Lots of extra space under the card after resizing

<img width="672" alt="Screen Shot 2021-05-25 at 12 57 34 PM" src="https://user-images.githubusercontent.com/2471874/119553326-caac6700-bd58-11eb-9b8b-7e9af030ca01.png">

I was at one point able to resize some cards such that they overlap - haven't trouble recreating this so maybe just a fluke, but potentially worth investigating

DeepakSap14 commented:
> Lots of extra space under the card after resizing
> 
> <img alt="Screen Shot 2021-05-25 at 12 57 34 PM" width="672" src="https://user-images.githubusercontent.com/2471874/119553326-caac6700-bd58-11eb-9b8b-7e9af030ca01.png">
> 
> I was at one point able to resize some cards such that they overlap - haven't trouble recreating this so maybe just a fluke, but potentially worth investigating

@mikerodonnell89 ,
i was able to re-produce this bug, i have fixed it now and pushed the change.

DeepakSap14 commented:
> <img alt="Screen Shot 2021-05-25 at 2 47 41 PM" width="1083" src="https://user-images.githubusercontent.com/39598672/119551958-30541f80-bd68-11eb-8935-5b60dd4095e6.png">
> 
> Shouldn't the layout be responsive? The cards should move to adjust the new screen size. Am I missing something?

Responsiveness on window resize is implemented, it can be re-reviewed.

DeepakSap14 commented:
> Why is the cursor set to `grab`?

i had handled cursor, but you asked to remove that here https://github.com/SAP/fundamental-ngx/pull/4964#discussion_r626842662.
i have handled the cursor again,
it is necessary in document side, because page is opening in pop-over and cursor changing accordingly.

InnaAtanasova commented:
> > Why is the cursor set to `grab`?
> 
> i had handled cursor, but you asked to remove that here [#4964 (comment)](https://github.com/SAP/fundamental-ngx/pull/4964#discussion_r626842662).
> i have handled the cursor again,
> it is necessary in document side, because page is opening in pop-over and cursor changing accordingly.

@DeepakSap14 let me try to explain the problem again. The Resizable card layout is a layout that uses cards (Card components) inside it. Noting of its functionality is related to the cursor. The card itself has 2 cursors: one that is a pointer, when you hover over the header, and the other one is a default one that is shown when you hover over the content. To be more precise, the second cursor should not be set at all as it depends on the content of the Card. All this means that if you have to modify the cursor in the documentation something's wrong somewhere in the code. Before changing the cursor in the documentation you had the `grab` cursor. A saw that you have some CDK classes applied. Why are they needed? By "fixing" stuff in the documentation we are just applying bandaids on more serious problems that might appear in the future.

"it is necessary in document side, because page is opening in pop-over and cursor changing accordingly."
That's not true. You are using Dialog component and Dialog component does not change the cursor to anything. Only draggable Dialog applies a grab cursor on the Dialog header, but none of them is changing the cursor over the body.

DeepakSap14 commented:
> > > Why is the cursor set to `grab`?
> > 
> > 
> > i had handled cursor, but you asked to remove that here [#4964 (comment)](https://github.com/SAP/fundamental-ngx/pull/4964#discussion_r626842662).
> > i have handled the cursor again,
> > it is necessary in document side, because page is opening in pop-over and cursor changing accordingly.
> 
> @DeepakSap14 let me try to explain the problem again. The Resizable card layout is a layout that uses cards (Card components) inside it. Noting of its functionality is related to the cursor. The card itself has 2 cursors: one that is a pointer, when you hover over the header, and the other one is a default one that is shown when you hover over the content. To be more precise, the second cursor should not be set at all as it depends on the content of the Card. All this means that if you have to modify the cursor in the documentation something's wrong somewhere in the code. Before changing the cursor in the documentation you had the `grab` cursor. A saw that you have some CDK classes applied. Why are they needed? By "fixing" stuff in the documentation we are just applying bandaids on more serious problems that might appear in the future.
> 
> "it is necessary in document side, because page is opening in pop-over and cursor changing accordingly."
> That's not true. You are using Dialog component and Dialog component does not change the cursor to anything. Only draggable Dialog applies a grab cursor on the Dialog header, but none of them is changing the cursor over the body.

Thanks @InnaAtanasova  for explanation.
i was assuming cursor change is default behaviour in dialog, which is wrong.
i see .cdk-drag present in utils/drag-and-drop is getting applied on fd-dialog-body. which causing this issue.

i will correct the example. will be using Dynamic page for creating example, which is better.

DeepakSap14 commented:
@InnaAtanasova 
I have changed my examples. now it is using dynamic page instead of dialog. Cursor issue is not coming here.
issue was not related to resizable-card component.

cursor issue is specific to dialog component. for re-producing, for eg: when you visit "Grid-list" component page and come back to "dialog" component, cursor changes to grab in "dialog component examples".
A issue can be raised for tracking this and fixing this in "dialog" component.

![2021-06-02_16-22-52 (1)](https://user-images.githubusercontent.com/57661487/120468969-696f3e00-c3bf-11eb-8260-bb76ed9c182e.gif)

N1XUS commented on apps/docs/src/app/core/component-docs/resizable-card-layout/examples/resizable-card-layout-example-itemconfig.component.ts:
```diff
@@ -0,0 +1,246 @@
+import { Component, OnInit } from '@angular/core';
+import { ResizableCardLayoutConfig, ResizedEvent } from '@fundamental-ngx/core';
```
Please import these from @fundamental-ngx/core/resizable-card-layout

DeepakSap14 replied:
fixed.

N1XUS commented on apps/docs/src/app/core/component-docs/resizable-card-layout/examples/resizable-card-layout-example-layoutconfig.component.ts:
```diff
@@ -0,0 +1,245 @@
+import { Component, OnInit } from '@angular/core';
+import { ResizableCardLayoutConfig, ResizedEvent } from '@fundamental-ngx/core';
```
Same as above

DeepakSap14 replied:
fixed.

N1XUS commented on libs/core/src/lib/resizable-card-layout/README.md:
```diff
@@ -0,0 +1,24 @@
+
+## Publishing
+
+After building your library with `ng build resizable-card-layout`, go to the dist folder `cd dist/resizable-card-layout` and run `npm publish`.
```
Dist path here is incorrect. I would suggest removing this content at all since it's a default angular documentation on how to build the lib

DeepakSap14 replied:
have removed publishing part for now.
